File: development/NET/samples/server/Console/VB/MyTransaction.vb
Imports System
Imports Softing.OPCToolbox
Imports Softing.OPCToolbox.Server

Namespace Console
	Public Class MyTransaction
		Inherits DaTransaction
#Region "Public Methods"
		'--------------------

		Public Sub New(ByVal aTransactionType As EnumTransactionType, ByVal requestList As DaRequest(), ByVal aSessionKey As UInt32)
			MyBase.New(aTransactionType, requestList, aSessionKey)
		End Sub

		Public Overloads Overrides Function HandleReadRequests() As Integer
			Dim count As Integer = Requests.Count
			Dim i As Integer = 0

			While i < count
				Dim request As DaRequest = DirectCast(Requests(i), DaRequest)

				If request.ProgressRequestState(EnumRequestState.CREATED, EnumRequestState.INPROGRESS) = True Then
					If request.PropertyId = 0 Then
						' get address space element value
						' take the Toolkit cache value
						Dim cacheValue As ValueQT = Nothing
						request.AddressSpaceElement.GetCacheValue(cacheValue)
						request.Value = cacheValue
						request.Result = EnumResultCode.S_OK
					Else
						' get property value
						' get the value from the address space element
						Dim element As MyDaAddressSpaceElement = CType(request.AddressSpaceElement, MyDaAddressSpaceElement)
						If Not element Is Nothing Then
							element.GetPropertyValue(request)
						Else
							request.Result = EnumResultCode.E_FAIL
						End If
					End If
				End If
				System.Math.Max(System.Threading.Interlocked.Increment(i), i - 1)
			End While
			Return CompleteRequests()
		End Function

		Public Overloads Overrides Function HandleWriteRequests() As Integer
			Dim count As Integer = Requests.Count

			Dim i As Integer = 0
			While i < count
				Dim request As DaRequest = CType(Requests(i), DaRequest)
				If Not request Is Nothing Then
					If request.ProgressRequestState(EnumRequestState.CREATED, EnumRequestState.INPROGRESS) = True Then

						Dim element As MyDaAddressSpaceElement = DirectCast(request.AddressSpaceElement, MyDaAddressSpaceElement)
						If Not element Is Nothing Then
							Dim valueQT As ValueQT = request.Value
							request.Result = element.ValueChanged(valueQT)
						End If
					End If
				End If
				System.Math.Max(System.Threading.Interlocked.Increment(i), i - 1)
			End While
			Return CompleteRequests()
		End Function

		'--
#End Region
	End Class
End Namespace

File: development/NET/samples/server/Console/CS/MyTransaction.cs
using Softing.OPCToolbox;
using Softing.OPCToolbox.Server;

namespace Console
{
	public class MyTransaction : DaTransaction
	{
		#region Public Methods
		//--------------------

		public MyTransaction(
			EnumTransactionType aTransactionType,
			DaRequest[] requestList,
			uint aSessionKey)
			:
			base(
			aTransactionType,
			requestList,
			aSessionKey)
		{
		}   //	end constructor

		public override int HandleReadRequests()
		{
			int count = Requests.Count;

			for (int i = 0; i < count; i++)
			{
				DaRequest request = (DaRequest)Requests[i];
				if ((request != null) && (request.ProgressRequestState(EnumRequestState.CREATED, EnumRequestState.INPROGRESS) == true))
				{
					if (request.PropertyId == 0)
					{
						// get address space element value
						// take the Toolkit cache value
						ValueQT cacheValue = null;
						request.AddressSpaceElement.GetCacheValue(ref cacheValue);
						request.Value = cacheValue;
						request.Result = EnumResultCode.S_OK;
					}
					else
					{
						// get property value
						// get the value from the address space element
						MyDaAddressSpaceElement element = request.AddressSpaceElement as MyDaAddressSpaceElement;
						if (element != null)
						{
							element.GetPropertyValue(request);
						}
						else
						{
							request.Result = EnumResultCode.E_FAIL;
						}   //	end if ... else
					}   //	end if ... else
				}   //	end if
			}   //	end for
			return CompleteRequests();
		}   //	end HandleReadRequests

		public override int HandleWriteRequests()
		{
			int count = Requests.Count;

			for (int i = 0; i < count; i++)
			{
				DaRequest request = Requests[i] as DaRequest;
				if (request != null)
				{
					MyDaAddressSpaceElement element = (MyDaAddressSpaceElement)request.AddressSpaceElement;
					if (element != null)
					{
						ValueQT valueQT = request.Value;
						request.Result = (EnumResultCode)element.ValueChanged(valueQT);
					}   //	end if
				}   //	end if
			}   //	end for
			return CompleteRequests();
		}   //	end HandleWriteRequests

		//--
		#endregion
	}   //	end class MyTransaction
}   //	end namespace
